From: Greg Banks <gnb@melbourne.sgi.com>
Date: Tue, 10 Dec 2013 20:24:48 -0600
Subject: [PATCH 2/3] Make NFSD DMAPI aware
References: bnc#859795 fate#315263
Patch-mainline: Never (dmapi was refused upstream)

G'day,

The NFSv3 protocol specifies an error, NFS3ERR_JUKEBOX, which a server
should return when an I/O operation will take a very long time.
This causes a different pattern of retries in clients, and avoids
a number of serious problems associated with I/Os which take longer
than an RPC timeout.  The Linux knfsd server has code to generate the
jukebox error and many NFS clients are known to have working code to
handle it.

One scenario in which a server should emit the JUKEBOX error is when
a file data which the client is attempting to access is managed by
an HSM (Hierarchical Storage Manager) and is not present on the disk
and needs to be brought in from tape.  Due to the nature of tapes this
operation can take minutes rather than the milliseconds normally seen
for local file data.

Currently the Linux knfsd handles this situation poorly.  A READ NFS
call will cause the nfsd thread handling it to block until the file
is available, without sending a reply to the NFS client.  After a
few seconds the client retries, and this second READ call causes
another nfsd to block behind the first one.  A few seconds later and
the client's retries have blocked *all* the nfsd threads, and all NFS
service from the server stops until the original file arrives on disk.

WRITEs and SETATTRs which truncate the file are marginally better, in
that the knfsd dupcache will catch the retries and drop them without
blocking an nfsd (the dupcache *will* catch the retries because the
cache entry remains in RC_INPROG state and is not reused until the
first call finishes).  However the first call still blocks, so given
WRITEs to enough offline files the server can still be locked up.

There are also client-side implications, depending on the client
implementation.  For example, on a Linux client an RPC retry loop uses
an RPC request slot, so reads from enough separate offline files can
lock up a mountpoint.

This patch seeks to remedy the interaction between knfsd and HSMs by
providing mechanisms to allow knfsd to tell an underlying filesystem
(which supports HSMs) not to block for reads, writes and truncates
of offline files.  It's a port of a Linux 2.4 patch used in SGI's
ProPack distro for the last 12 months.  The patch:

*  provides a new ATTR_NO_BLOCK flag which the kernel can
   use to tell a filesystem's inode_ops->setattr() operation not
   to block when truncating an offline file.  XFS already obeys
   this flag (inside a #ifdef)

*  changes knfsd to provide ATTR_NO_BLOCK when it does the VFS
   calls to implement the SETATTR NFS call.

*  changes knfsd to supply the O_NONBLOCK flag in the temporary
   struct file it uses for VFS reads and writes, in order to ask
   the filesystem not to block when reading or writing an offline
   file.  XFS already obeys this new semantic for O_NONBLOCK
   (and in SLES9 so does JFS).

*  adds code to translate the -EAGAIN the filesystem returns when
   it would have blocked, to the -ETIMEDOUT that knfsd expects.

Edits by jeffm against Linux v4.4:
* Upstream commit 062304a81 (nfsd: stop translating EAGAIN to nfserr_dropit)
  removed the need to remap EAGAIN.
* Moved the XFS portion into the XFS patch.

[js] ATTR_NO_BLOCK shall be renumbered to something like (1 << 30) to avoid
     collisions with new upstream flags.

Signed-off-by: Greg Banks <gnb@melbourne.sgi.com>
(SLES9 patch Acked-by: okir@suse.de)
Signed-off-by: NeilBrown <neilb@suse.de>
Acked-by: Jan Kara <jack@suse.cz>

Conflicts:
	fs/xfs/linux-2.6/xfs_iops.c
---

 fs/nfsd/vfs.c      |   17 +++++++++++++++++
 include/linux/fs.h |    3 ++-
 2 files changed, 19 insertions(+), 1 deletion(-)

--- a/fs/nfsd/vfs.c
+++ b/fs/nfsd/vfs.c
@@ -379,6 +379,15 @@ nfsd_setattr(struct svc_rqst *rqstp, str
 	/* Callers that do fh_verify should do the fh_want_write: */
 	get_write_count = !fhp->fh_dentry;
 
+	/*
+	 * Tell a Hierarchical Storage Manager (e.g. via DMAPI) to
+	 * return EAGAIN when an action would take minutes instead of
+	 * milliseconds so that NFS can reply to the client with
+	 * NFSERR_JUKEBOX instead of blocking an nfsd thread.
+	 */
+	if (rqstp->rq_vers >= 3)
+		iap->ia_valid |= ATTR_NO_BLOCK;
+
 	/* Get inode */
 	err = fh_verify(rqstp, fhp, ftype, accmode);
 	if (err)
@@ -864,6 +873,10 @@ static __be32
 nfsd_vfs_read(struct svc_rqst *rqstp, struct file *file,
 	      loff_t offset, struct kvec *vec, int vlen, unsigned long *count)
 {
+	/* Support HSMs -- see comment in nfsd_setattr() */
+	if (rqstp->rq_vers >= 3)
+		file->f_flags |= O_NONBLOCK;
+
 	if (file->f_op->splice_read && test_bit(RQ_SPLICE_OK, &rqstp->rq_flags))
 		return nfsd_splice_read(rqstp, file, offset, count);
 	else
@@ -940,6 +953,10 @@ nfsd_vfs_write(struct svc_rqst *rqstp, s
 	if (!EX_ISSYNC(exp))
 		stable = 0;
 
+	/* Support HSMs -- see comment in nfsd_setattr() */
+	if (rqstp->rq_vers >= 3)
+		file->f_flags |= O_NONBLOCK;
+
 	/* Write the data. */
 	oldfs = get_fs(); set_fs(KERNEL_DS);
 	host_err = vfs_writev(file, (struct iovec __user *)vec, vlen, &pos);
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -226,7 +226,8 @@ typedef void (dax_iodone_t)(struct buffe
 #define ATTR_KILL_PRIV	(1 << 14)
 #define ATTR_OPEN	(1 << 15) /* Truncating from open(O_TRUNC) */
 #define ATTR_TIMES_SET	(1 << 16)
-#define ATTR_TOUCH	(1 << 17)
+#define ATTR_NO_BLOCK	(1 << 17)
+#define ATTR_TOUCH	(1 << 18) /* kABI, renumber ATTR_NO_BLOCK to 30 in SP3 */
 
 /*
  * Whiteout is represented by a char device.  The following constants define the
